documentation/internal/Development-Workflow.md

```diff
@@ -0,0 +1,311 @@
+
+_Why do we update the cache after mutations?_
+
+When we create a new entity that has a relationshio with some existing in cache we need to also update the cache with the added entity! This is to avoid having to create a `refetch` to get this back from the server and keep the local cache up-to-date. The same way, after editing an entity which is already in cache we need to remember to update this entity in the cache.
+
+##### Example of remote data
+
+- Current Application
+- Application Sections (related enetities - Application)
+  - Template Sections (related enetities - Application Section)
+- Application Responses (related entities - Application)
+  - Template element (related entities - Application Response)
+
+### Local states
+
+We are using 2 options for local states:
+
+- local variables (with `useState`) passed as `props` to other components
+- context areas (with `useContext`) for sharing between components
+
+##### Example of local state
+
+**TODO**: Rename ApplicationPage => ApplicationPageWrapper (if we decide that we want to keep the wrapper thingy).
+
+**To be created** => In ApplicationPageWrapper:
+
+- List of all questions and responses
+
+How to use:
+
+1. `ApplicationPage` calls initial query to load the whole application, section and responses.
+2. `ApplicationPage` calls a custom hook to create the list of questions and responses based on query.
+3. Custom hook returns the result object Q&R with all the question elements with reduced fields (after running the `evaluateExpression` on each question) and the responses by the user.
+
+The object returned has for each Q&R (Question and Response):
+
+- Question code (key)
+- Type: Information | Question - if information none of the next fields are filled!
+- isRequired: boolean
+- isVisible: boolean
+- isValid: boolean
+- ResponseId: number
+- ResponseJSON: JSON
+- ResponseValue: string
+
+4. The Q&R object is passed to `ElementsBox` and each related question and response to the `ApplicationViewWrapper`.
+5. The `ApplicationViewWrapper` receives the mutation (from `ElementsBox`) to update the response `onFieldChage`. It should only update to the server and CACHE when the response is valid.
+6. The `ApplicationPage` will update the Q&R object when the query for elements and responses gets triggerdby the new mutation of one of the elements response on the current application.
+
+##### Example of context state
+
+`src/contexts/ApplicationState.tsx`
+
+This context area is used between components to share a few states:
+
+- Application serial number (currently created on Front-end)
+- Application id
+- Application's sections total of pages - which gets deduced by the elements in each section.
+
+---
+
+# Development workflow
+
+## Example
+
+Process of creating one Application, going through each page until reach the summary page where the user can submit.
+
+![development_workflow_applications](images/Development_workflow_applications.png)
```

Comment:
Image is great. Just one thing: I spent ages wondering what on earth "bt" was. Then I realise you meant "button". If there's no room for the full word "button", maybe just lose the "bt". It's not an abbreviation I've seen before and the diagram box looks enough like an actual button to get the idea.

Reply:
Hahhaa Sorry! I'll just drop it then : D